### PR TITLE
Add Pyramid 2.0 compatibility

### LINKED ISSUE #32: Incompatible with Pyramid 2.0
The module pyramid.compat has been removed in 2.0. pyramid_ldap has to be changed accordingly.

```
  File "/home/marcel/.pyenv/versions/3.9.2/envs/dash/lib/python3.9/site-packages/pyramid/config/__init__.py", line 632, in include
    c = self.maybe_dotted(callable)
  File "/home/marcel/.pyenv/versions/3.9.2/envs/dash/lib/python3.9/site-packages/pyramid/config/__init__.py", line 740, in maybe_dotted
    return self.name_resolver.maybe_resolve(dotted)
  File "/home/marcel/.pyenv/versions/3.9.2/envs/dash/lib/python3.9/site-packages/pyramid/path.py", line 327, in maybe_resolve
    return self._resolve(dotted, package)
  File "/home/marcel/.pyenv/versions/3.9.2/envs/dash/lib/python3.9/site-packages/pyramid/path.py", line 334, in _resolve
    return self._zope_dottedname_style(dotted, package)
  File "/home/marcel/.pyenv/versions/3.9.2/envs/dash/lib/python3.9/site-packages/pyramid/path.py", line 383, in _zope_dottedname_style
    found = __import__(used)
  File "/home/marcel/.pyenv/versions/3.9.2/envs/dash/lib/python3.9/site-packages/pyramid_ldap/__init__.py", line 19, in <module>
    from pyramid.compat import (
ModuleNotFoundError: No module named 'pyramid.compat'
```

Also, set_request_property is no longer available:

```
  File "/home/marcel/.pyenv/versions/3.9.2/envs/dash/lib/python3.9/site-packages/pyramid/config/actions.py", line 573, in wrapper
    result = wrapped(self, *arg, **kw)
  File "/home/marcel/.pyenv/versions/3.9.2/envs/dash/lib/python3.9/site-packages/pyramid_ldap/__init__.py", line 304, in ldap_setup
    config.set_request_property(get_connector, 'ldap_connector', reify=True)
  File "/home/marcel/.pyenv/versions/3.9.2/envs/dash/lib/python3.9/site-packages/pyramid/config/__init__.py", line 716, in __getattr__
    raise AttributeError(name)
AttributeError: set_request_property
```

### PR DESCRIPTION
**Pull request by msander**
- replace pyramid.compat by six
- replace set_request_property by add_request_method
- fix flake8 issues

fixes #32, #26 and #28

### CODE REVIEW

**Comment by bertjwregeer** [comment]
Pyramid 2.0 is strictly Python 3, replacing `pyramid.compat` with `six` seems a bit silly. In most cases we can rely on `b` prefixes where necessary.

File: pyramid_ldap/__init__.py [hunk 1]
```diff
@@ -129,13 +129,13 @@ def authenticate(self, login_unsafe, password_unsafe):
         :exc:`pyramid.exceptions.ConfiguratorError`.
 
         In pyramid_ldap <= version 0.2, authenticating with a login that
-        included the domain (e.g. CORP\exampleuser) would raise
-        ``ldap.FILTER_ERROR`` because the ``\`` led to an invalid search
+        included the domain (e.g. CORP\\exampleuser) would raise
```
Comment (stevepiercy) [inline pyramid_ldap/__init__.py:132]: This should be a literal.
```suggestion
        included the domain (e.g., ``CORP\exampleuser``) would raise
```

File: pyramid_ldap/__init__.py [hunk 2]
```diff
@@ -129,13 +129,13 @@ def authenticate(self, login_unsafe, password_unsafe):
...
+        ``ldap.FILTER_ERROR`` because the ``\\`` led to an invalid search
```
Comment (stevepiercy) [inline pyramid_ldap/__init__.py:133]: Please revert.
```suggestion
        ``ldap.FILTER_ERROR`` because the ``\`` led to an invalid search
```

File: pyramid_ldap/__init__.py [hunk 3]
```diff
@@ -129,13 +129,13 @@ def authenticate(self, login_unsafe, password_unsafe):
...
         string. In pyramid_ldap >= 0.3, the string is escaped so it will not
         raise an exception. However, it will likely fail to authenticate user
         ``CORP\\5cexampleuser`` (the escaped form of login
-        "CORP\exampleuser"). Applications using pyramid_ldap can preprocess the
-        logins to make sure they are formatted correctly for their
+        "CORP\\exampleuser"). Applications using pyramid_ldap can preprocess
```
Comment (stevepiercy) [inline pyramid_ldap/__init__.py:137]: This should also be a literal.

```suggestion
        ``CORP\exampleuser``). Applications using pyramid_ldap can preprocess
```

**Review by stevepiercy — CHANGES_REQUESTED** [review]

**Comment by stevepiercy** [comment]
@bertjwregeer should we drop EOLed Pythons and add supported versions, as well as move this to GitHub Actions?

**Comment by bertjwregeer** [comment]
Yes, this extension needs a lot of love as it has received none over the years, but just dropping Py2 support and matching Pyramid 2.0 would be a good start.

**Comment by stevepiercy** [comment]
@bertjwregeer how should we proceed? We could do it all in one swoop—fixing bugs, dropping EOLed Pythons, moving to GHA, and adding Pyramid 2.0 compat—or do it in baby steps to cut two or more releases. I can do the work, just need advice and direction.

**Comment by msander** [comment]
> Pyramid 2.0 is strictly Python 3, replacing `pyramid.compat` with `six` seems a bit silly. In most cases we can rely on `b` prefixes where necessary.

Pyramid 2.0 compatibility does not mean that you need to drop Pyramid 1.x compatibility (which supports Python 2). 
However, Python 2 is EOL and I have not touched any Python 2 projects in years. So I guess it would be a reasonable decision.
If you want, I can modify my pull request and remove `six`.

**Review by stevepiercy — APPROVED** [review]
Thank you!

**Comment by msander** [comment]
> Thank you for the docstring changes!

Thank you for you review.

However, PEP8 is failing again due to these changes:

```
pyramid_ldap/__init__.py:132:42: W605 invalid escape sequence '\e'
pyramid_ldap/__init__.py:133:45: W605 invalid escape sequence '\`'
pyramid_ldap/__init__.py:137:15: W605 invalid escape sequence '\e'
```
Don't we need double backslashes here?

**Comment by stevepiercy** [comment]
Flake8 gives a false positive. This is a reStructuredText inline literal, but Flake8 does not interpret it as such. See screenshot for how it renders to HTML with the changes I requested and which you applied.

Screen Shot 2021-05-04 at 3 15 42 AM

I could not find in PEP8 that it covers this edge case. To avoid the Flake8 warning, change the last line of the docstring to ignore this warning:

```
"""  # noqa: W605
```

Also I missed a couple of changes in `tests.py` that should have the same `  # noqa: W605`:

```
pyramid_ldap/tests.py:206:31: W605 invalid escape sequence '\l'
pyramid_ldap/tests.py:216:40: W605 invalid escape sequence '\*'
```

...would avoid the warning, pass tests, and not obscure the original intent of the test:

```python
        inst.authenticate('BAD\login', 'password')  # noqa: W605
#...
        inst.authenticate('login', 'bad\*()password')  # noqa: W605
```

**Comment by msander** [comment]
Thank you for your explanation. 
Nevertheless, I think the double backslash is the correct solution.
The alternative would be to use a raw string (r"...").
An invalid escape like "\*" is currently interpreted as "\\*", but will result in a SyntaxError in the future. That's what the warning is about.

https://www.flake8rules.com/rules/W605.html

Also the tests are correct and finally pass with the double backslash.

The documentation is also rendered the same.

**Comment by stevepiercy** [comment]
OK, thank you for explaining. I misunderstood. I apologize for my mistake.

We have two failed build jobs, one of which is an allowed failure for Python 3.7 and the other for a false positive that coverage is below 100%. We can skip the Python 3.7 failure for now, but we need to resolve the coverage issue. I've seen this issue in other repos. I'll look for how we fixed it. [It was working just prior to this PR](https://travis-ci.com/github/Pylons/pyramid_ldap/builds/197640295).

**Comment by stevepiercy** [comment]
When I run `tox -r -e py2-cover,py3-cover,coverage` locally, it passes. I checked closely in the Travis log, and noticed ["Cache entry deserialization failed, entry ignored"](https://travis-ci.com/github/Pylons/pyramid_ldap/jobs/502857458#L248). I [cleared the cache](https://docs.travis-ci.com/user/caching#clearing-caches), and restarted the build. It still fails. I have no clue how to fix this, other than try moving it to GitHub Actions and forget about Travis.

**Comment by msander** [comment]
Tests are passing now.

@bertjwregeer in setup.py, there is a comment ```'six',  # required by `ldappool` but not in their requirements file```
This seems still to be true and I think we should keep that requirement. Do you still want to remove the usage?

File: tox.ini [hunk 4]
```diff
@@ -40,7 +40,7 @@ setenv =
 [testenv:py3-cover]
 basepython = python3.6
 commands =
-    pip install -q pyramid[testing]
+    pip install -q pyramid_ldap[testing]
```
Comment (stevepiercy) [inline tox.ini:43]: Nice catch!

**Review by stevepiercy — APPROVED** [review]

**Comment by stevepiercy** [comment]
> Tests are passing now.
> 
> @bertjwregeer in setup.py, there is a comment `` 'six', # required by `ldappool` but not in their requirements file ``
> This seems still to be true and I think we should keep that requirement. Do you still want to remove the usage?

I see [no instances of `six` in its code](https://opendev.org/openstack/ldappool/search?q=six), so maybe it is OK to remove? Please try and let's see what happens.

**Comment by msander** [comment]
Please take a look at
https://opendev.org/openstack/ldappool/src/branch/master/ldappool/__init__.py

File: setup.py [hunk 5]
```diff
@@ -13,7 +13,7 @@
 
 requires = [
     'pyramid>=1.3',
-    'six', # required by `ldappool` but not in their requirements file
+    'six',  # required by `ldappool` but not in their requirements file
```
Comment (bertjwregeer) [inline setup.py:16]: If this is no longer required because `ldappool` itself is Py3 only and no longer uses `six` then let's remove this.

**Comment by bertjwregeer** [comment]
> Please take a look at
> https://opendev.org/openstack/ldappool/src/branch/master/ldappool/__init__.py

I missed this... so then yes, keep the requirement for `six`.

**Comment by stevepiercy** [comment]
@bertjwregeer I think this is OK to merge, unless you have any further review.

I can work on bringing it to Pylons Project standards and parity with Pyramid 2.0 for Python support and GHA later this week.
